models/points.go

```diff
@@ -1386,70 +1391,169 @@ func (p *point) UnixNano() int64 {
+func (a Tags) GetString(key string) string {
+	return string(a.Get([]byte(key)))
+}
+
+// Set sets the value for a key.
+func (a *Tags) Set(key, value []byte) {
+	for _, t := range *a {
+		if bytes.Equal(t.Key, key) {
+			t.Value = value
+			return
+		}
+	}
+	*a = append(*a, Tag{Key: key, Value: value})
+	sort.Sort(*a)
+}
+
+// SetString sets the string value for a string key.
+func (a *Tags) SetString(key, value string) {
+	a.Set([]byte(key), []byte(value))
+}
+
+// Delete removes a tag by key.
+func (a *Tags) Delete(key []byte) {
+	for i, t := range *a {
+		if bytes.Equal(t.Key, key) {
+			copy((*a)[i:], (*a)[i+1:])
+			(*a)[len(*a)-1] = Tag{}
+			*a = (*a)[:len(*a)-1]
+			return
+		}
+	}
+}
+
+// Map returns a map representation of the tags.
+func (a Tags) Map() map[string]string {
+	m := make(map[string]string, len(a))
+	for _, t := range a {
+		m[string(t.Key)] = string(t.Value)
+	}
+	return m
+}
 
 // Merge merges the tags combining the two. If both define a tag with the
 // same key, the merged value overwrites the old value.
 // A new map is returned.
-func (t Tags) Merge(other map[string]string) Tags {
-	merged := make(map[string]string, len(t)+len(other))
-	for k, v := range t {
-		merged[k] = v
+func (a Tags) Merge(other map[string]string) Tags {
+	merged := make(map[string]string, len(a)+len(other))
+	for _, t := range a {
+		merged[string(t.Key)] = string(t.Value)
 	}
 	for k, v := range other {
 		merged[k] = v
 	}
-	return Tags(merged)
+	return NewTags(merged)
 }
 
 // HashKey hashes all of a tag's keys.
-func (t Tags) HashKey() []byte {
+func (a Tags) HashKey() []byte {
 	// Empty maps marshal to empty bytes.
-	if len(t) == 0 {
+	if len(a) == 0 {
 		return nil
 	}
 
-	escaped := Tags{}
-	for k, v := range t {
-		ek := escapeTag([]byte(k))
-		ev := escapeTag([]byte(v))
+	escaped := make(Tags, 0, len(a))
+	for _, t := range a {
+		ek := escapeTag(t.Key)
+		ev := escapeTag(t.Value)
 
 		if len(ev) > 0 {
-			escaped[string(ek)] = string(ev)
+			escaped = append(escaped, Tag{Key: ek, Value: ev})
 		}
 	}
 
 	// Extract keys and determine final size.
 	sz := len(escaped) + (len(escaped) * 2) // separators
-	keys := make([]string, len(escaped)+1)
-	i := 0
-	for k, v := range escaped {
-		keys[i] = k
-		i++
-		sz += len(k) + len(v)
+	keys := make([][]byte, len(escaped)+1)
+	for i, t := range escaped {
+		keys[i] = t.Key
+		sz += len(t.Key) + len(t.Value)
 	}
-	keys = keys[:i]
-	sort.Strings(keys)
+	keys = keys[:len(escaped)]
+	sort.Sort(byteSlices(keys))
+
 	// Generate marshaled bytes.
 	b := make([]byte, sz)
 	buf := b
 	idx := 0
-	for _, k := range keys {
+	for i, k := range keys {
 		buf[idx] = ','
 		idx++
 		copy(buf[idx:idx+len(k)], k)
 		idx += len(k)
 		buf[idx] = '='
 		idx++
-		v := escaped[k]
+		v := escaped[i].Value
 		copy(buf[idx:idx+len(v)], v)
 		idx += len(v)
 	}
 	return b[:idx]
 }
 
+// Dereference removes all references to data within b and moves them to the heap.
+func (a Tags) Dereference(b []byte) {
```

Comment:
I don't like that this logic exists on the Tags type itself. Since all the fields are exported, I would rather see this moved to where we're already using unsafe/mmap.